### ISSUE TITLE
Screen Sharing with like 5FPS and Screen Sharing not playing when in fullscreen[Bug]

### ISSUE THREAD

**Issue by Daki0082** [SpikeHD/Dorion#153]
**Describe the bug**
Every time I share my screen on Dorion, my friends say that my game runs at 5 FPS. And when my game is in full screen, my friends only see a still image of my game. They only see my game moving when I tab out of the fullscreen game, again at 5FPS. This only happens on Dorion, not on the normal Discord app.

**To Reproduce**
- Open Dorion
- Call a friend or join a voice channel
- share your screen

**System Info (please complete the following information):**
 - OS: Windows 11
 - Version: happens on all Dorion versions

**Comment by SpikeHD** [comment]
Looks like this might be due to [this bug](https://bugs.chromium.org/p/webrtc/issues/detail?id=13016) with the VP9 codec. I think I can experiment with forcing the VP8 codec but no promises 😄

**Comment by Daki0082** [comment]
Okay, thanks!

**Comment by SpikeHD** [comment]
@Daki0082 Okay, after some investigation, this seems to be more of a "thing" than I thought, so if you're willing, I'd like to get you to answer/try a couple things for me:

1. What is your setup like? Are you using multi-monitors? Are you using 1 monitor connected to your motherboard, and one connected to your dedicated GPU? Do you even have a dedicated GPU?
2. Open the DevTools console in Dorion (`Ctrl+Shift+i`) and paste and run the following code (yea yea I know this is sketch, you don't have to if you don't want to). This will output your GPU details, I want to make sure whatever hardware you're running is getting detected properly since Dorion *should* use hardware acceleration. Just paste the output back to me:
```js
var canvas = document.createElement('canvas');
var gl;
var debugInfo;
var vendor;
var renderer;

try {
  gl = canvas.getContext('webgl') || canvas.getContext('experimental-webgl');
} catch (e) {}

if (gl) {
  debugInfo = gl.getExtension('WEBGL_debug_renderer_info');
  vendor = gl.getParameter(debugInfo.UNMASKED_VENDOR_WEBGL);
  renderer = gl.getParameter(debugInfo.UNMASKED_RENDERER_WEBGL);
}

console.log(vendor)
console.log(renderer)
```
Mine outputted:
```
Google Inc. (NVIDIA)
VM692:20 ANGLE (NVIDIA, NVIDIA GeForce RTX 3070 (0x00002484) Direct3D11 vs_5_0 ps_5_0, D3D11)
```

3. If you have a friend willing to put in some time (NOT just using an alt locally, that won't give accurate data. trust me, I've tried, lol), I'd like you to look at the following:
  a. You, in **Dorion**, start a screenshare like normal, ideally of a game or something you know tanks the FPS
  b. Your friend, in a normal client, opens the stream. Then, they click the green "Voice Connected" text, and on the bottom of that popup, they should see "Debug". Click that. If it doesn't show up, [enable debug mode](https://www.howtogeek.com/714348/how-to-enable-or-disable-developer-mode-on-discord/) in Discord and try again.
   ![image](https://github.com/SpikeHD/Dorion/assets/25207995/c8a85a9a-ad65-471b-af38-55bdb464d9d0)
  c. In the box that should now appear, they click on the user (you) that is streaming (under "RTC Debug: Media"), which should open something like below. Make sure it is the "Video" debug tab, not "Audio". **Screenshot that and send it here as well**, below is what you're looking for:![image](https://github.com/SpikeHD/Dorion/assets/25207995/b0a75c58-3146-4e39-8dca-848509ed214c)

I will also hopefully remember to at least add an option for disabling hardware acceleration in the next release which may or may not help, if I am unable to fix this before the next one.

For any nerds reading, my thoughts:
1. Since WebRTC is P2P, it is interesting that when reproducing locally (Dorion streaming and browser viewing, on the same PC), this problem seems to not exist. This suggests it could be a network-related problem?
2. Screensharing also loves to crank the CPU like soulja boy. I'm noticing in Task Manager that screensharing uses GPU 0 (Intel HD). Unsure why that's the case when Dorion detects my GPU just fine, but it may be because I use a weird setup that uses the iGPU for one monitor, and my 3070 for the other?

**Comment by SpikeHD** [comment]
Tested with a friend as well, my findings:

1. Browser screensharing good. Closer to 30fps. Bitrate was fine.
2. Dorion screensharing consistently bad. Never faltered from low FPS (5-ish). Bitrate was fine.
3. Browser webcam consistently good. Close to target FPS. Bitrate and resolution matched target.
4. Dorion webcam started bad, then gradually improved. Started at low FPS, went up to close to target. Bitrate started terrible as well, ended about the same as browser.

All-in-all, I think something is up with WebView2s implementation of screensharing, as webcam seemed to work fine. Going to look into this further but you are still welcome to answer the stuff above.

**Comment by SpikeHD** [comment]
@Daki0082 Another discovery: streaming a full desktop instead of a window will make the framerate totally fine. You can also share audio when streaming entire desktops instead of Windows. Give that a try, see if that is the case for you as well.

**Comment by SpikeHD** [comment]
Created an upstream issue in WebView2Feedback: https://github.com/MicrosoftEdge/WebView2Feedback/issues/4176

**Comment by Daki0082** [comment]
> @Daki0082 Okay, after some investigation, this seems to be more of a "thing" than I thought, so if you're willing, I'd like to get you to answer/try a couple things for me:
> 
> 1. What is your setup like? Are you using multi-monitors? Are you using 1 monitor connected to your motherboard, and one connected to your dedicated GPU? Do you even have a dedicated GPU?
> 2. Open the DevTools console in Dorion (`Ctrl+Shift+i`) and paste and run the following code (yea yea I know this is sketch, you don't have to if you don't want to). This will output your GPU details, I want to make sure whatever hardware you're running is getting detected properly since Dorion _should_ use hardware acceleration. Just paste the output back to me:
> 
> ```js
> var canvas = document.createElement('canvas');
> var gl;
> var debugInfo;
> var vendor;
> var renderer;
> 
> try {
>   gl = canvas.getContext('webgl') || canvas.getContext('experimental-webgl');
> } catch (e) {}
> 
> if (gl) {
>   debugInfo = gl.getExtension('WEBGL_debug_renderer_info');
>   vendor = gl.getParameter(debugInfo.UNMASKED_VENDOR_WEBGL);
>   renderer = gl.getParameter(debugInfo.UNMASKED_RENDERER_WEBGL);
> }
> 
> console.log(vendor)
> console.log(renderer)
> ```
> 
> Mine outputted:
> 
> ```
> Google Inc. (NVIDIA)
> VM692:20 ANGLE (NVIDIA, NVIDIA GeForce RTX 3070 (0x00002484) Direct3D11 vs_5_0 ps_5_0, D3D11)
> ```
> 

This is what mine outputted: 

```
Google Inc. (NVIDIA)
ANGLE (NVIDIA, NVIDIA GeForce GTX 1650 (0x00001F0A) Direct3D11 vs_5_0 ps_5_0, D3D11)
```


I'm going to ask a friend of mine to help me soon. I will respond again with the screenshot when I have it

**Comment by Daki0082** [comment]
Here's the screenshot of the debugging thing: 
![image](https://github.com/SpikeHD/Dorion/assets/87087409/ab249529-4a95-44ae-9dff-bfc18cf00374)

**Comment by SpikeHD** [comment]
Okay, thank you, it seems it is what I feared most: a bug with the underlying browser engine :(

I can't fix this myself, but I did submit a (hopefully) detailed issue in WebView2Feedback, so hopefully that leads to something. In the meantime, you can try screensharing your entire desktop instead of specific windows, since I'm pretty sure the bug is present in the window capturing system. At least, sharing my desktop works fine for me, feel free to report back if it doesn't for you 😄

**Comment by SpikeHD** [comment]
Yep, nothing I can do until the [upstream issue](https://github.com/MicrosoftEdge/WebView2Feedback/issues/4176) gets an update!

**Comment by noideas4thisname** [comment]
Yep, definitely encountered this. Makes Dorion unusable for this use care at the moment. Unfortunately upstream seems to not have any traction that we can see.

**Comment by SpikeHD** [comment]
I've experimented with the Vencord [webScreenShareFixes](https://github.com/Vendicated/Vencord/blob/main/src/plugins/webScreenShareFixes.web/index.ts) plugin, and it seemed to bring the bitrate from ~2500 Kbps to ~3500 Kbps but in more complex scenarios (I tested with Overwatch practice range) it still has very poor FPS.

**Comment by MrJuaumBR** [comment]
So i use Windows 11, only one monitor, stay the same issue :(
Also i'm using this plugin from vencord(webScreenShareFixes)

<img width="571" height="56" alt="Image" src="https://github.com/user-attachments/assets/5fea4fcc-5ac5-45d4-bea2-6fedb512acf9" />

<img width="738" height="1026" alt="Image" src="https://github.com/user-attachments/assets/f8bc51f3-40d7-4c12-8bb3-aa027142968c" />

**Comment by craciu25yt** [comment]
not sure if this stills under investigation but I've seen that discord's encoder is nvidia directx3d and webview/dorion use openh264. not sure if that has anything to do but yea while moving on games such as cs2 the frames drop to 6. Also seems like they drop the most when more movement is happening but no bitrate threshold is happening as it goes above the 2500kbps cap with the vencord plugin.

edit: 
also as full screen is even worst

<img width="832" height="985" alt="Image" src="https://github.com/user-attachments/assets/b797de55-ec9f-45cf-bbae-7b27fd1eea4c" />
